### PR TITLE
DApp-1549 Dropdown/Menu component

### PR DESCRIPTION
**Pull request by corlard3y**
## Pull Request Template

#1545 

- [#1545](link-to-ticket)

### Description

- Introduce menu in blocks folder which has Menu component, which serves as parent component to the Menu Item component
- Added MenuItem component, which is the list of all items in the Menu component. 
- Added Dropdown component too, which has a trigger to open the Menu(overlay), either on Hover action, or onClick action.
- Also, added light and dark mode support to these components 

### Problem/Feature
Create a dropdown and menu package in the blocks folder and introduce these dropdown and menu items as a react component so that these could be reused in other component.

Figma: https://www.figma.com/design/FvJL0JWHlsVohsHTERB1Lh/Push-Design-Foundations-%5BFuture%5D?node-id=2663-23523&t=XRHuzgCLtQkYnb4h-1

### Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Documentation update
- [ ] Other (please describe):

### Checklist

- [ ] **Quick PR**: Is this a quick PR? Can be approved before finishing a coffee.
  - [ ] Quick PR label added
- [ ] **Not Merge Ready**: Is this PR dependent on some other PR/tasks and not ready to be merged right now.
  - [ ] DO NOT Merge PR label added

### Frontend Guidelines

- [ ] Followed frontend guidelines as per the [Guidelines Notion Doc](https://www.notion.so/pushprotocol/Frontend-dApp-Guidelines-1d7806ae3d9e4569a340b563dcd0536c)

### Build & Testing

- [x] No errors in the build terminal
- [ ] Engineer has tested the changes on their local environment
- [ ] Engineer has tested the changes on deploy preview

### Screenshots/Video with Explanation

- **Before:** Explain the previous behavior

- **After:** What's changed now

### Review & Approvals

- [ ] Self-review completed
- [ ] Code review by at least one other engineer
- [ ] Documentation updates if applicable

### CODE REVIEW

**Comment by corlard3y** [comment]
Added the basic flow of the dropdown and Menu, Pls lmk what changes can be added. Also, used hardcoding within the dropdown component so I will need your assistance/direction within the children(isOpen) param, and title ig @rohitmalhotra1420

File: src/blocks/dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
+import { ReactNode, forwardRef, useState } from 'react';
+import styled, { FlattenSimpleInterpolation } from 'styled-components';
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:2]:
```suggestion
import { forwardRef, useState } from 'react';
import styled from 'styled-components';
```
remove unused imports

File: src/blocks/dropdown/Dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      case 'bottomRight':
+        return 'bottomRight';
+      default:
+        return 'bottom';
+    }
+  };
+
+  return (
+    <StyledDropdown
+      ref={ref}
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:110]:
```suggestion
      ref={ref}
```

Fix the typescript error

File: src/blocks/dropdown/Dropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
+import { ReactNode, forwardRef, useState } from 'react';
+import styled, { FlattenSimpleInterpolation } from 'styled-components';
+
+import { getBlocksColor } from 'blocks/Blocks.utils';
+import { useBlocksTheme } from 'blocks/Blocks.hooks';
+import { ModeProp } from 'blocks/Blocks.types';
+import { DropdownProps } from './Dropdown.types';
+
+const StyledDropdown = styled.div.withConfig({
+  shouldForwardProp: (prop, defaultValidatorFn) => !['mode'].includes(prop) && defaultValidatorFn(prop),
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:10]: Add all the props in this array which are passed down to the html tags

File: src/blocks/dropdown/Dropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+})<DropdownProps & ModeProp>`
+  position: relative; /* Ensure the dropdown has relative positioning */
+  min-width: 200px;
+  .dropdown-item {
+    background-color: ${({ mode }) => getBlocksColor(mode, { light: 'white', dark: 'gray-900' })};
+    border: 1px solid ${({ mode }) => getBlocksColor(mode, { light: 'gray-200', dark: 'gray-800' })};
+    padding: 10px 15px;
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:17]:
```suggestion
    padding: 10px 15px;
```
use `var(--s0)` format for gap, padding and margins

File: src/blocks/dropdown/Dropdown.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    border-radius: 20px;
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:18]:
```suggestion
    border-radius: 20px;
```
use var(--r0) format for border-radius

File: src/blocks/dropdown/Dropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    max-width: fit-content;
+    box-sizing: border-box;
+    // position: relative;
+    &:hover {
+      cursor: pointer;
+    }
+    span {
+      font-size: 15px;
+      font-weight: 400;
+      color: #657795;
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:32]: use var(--pink-100) format for colors

File: src/blocks/dropdown/Dropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    }
+  }
+  .menu-overlay {
+    position: absolute;
+    border-radius: 4px;
+    margin-top: 3px;
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:38]: Same

File: src/blocks/dropdown/Dropdown.tsx (hunk 4 above)
Comment (rohitmalhotra1420) [reply src/blocks/dropdown/Dropdown.tsx:17]: @corlard3y only the css variables mentioned App.tsx are to be allowed.

File: src/blocks/dropdown/Dropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      mode={mode}
+      trigger={trigger}
+      placement={placement}
+      onMouseEnter={handleMouseEnter}
+      onMouseLeave={handleMouseLeave}
+      {...props}
+    >
+      <div className='dropdown-item' onClick={handleClick}>
+        <span>Actions</span>
+      </div>
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:120]: Why do we need this actions button? the overlay should be visible when the trigger element is clicked. Right now the overlay is visible when the Actions text is clicked.

Screenshot 2024-06-20 at 11 35 18 AM

File: src/blocks/dropdown/Dropdown.types.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { ReactNode } from 'react';
+
+export type DropdownProps = {
+  // This will be content upon clicking on which the dropdown overlay will open
+  children?: (isOpen: boolean) => ReactNode;
+  placement?: 'bottom' | 'bottomLeft' | 'top' | 'topRight' | 'bottomRight' | 'topLeft';
+  // on which action to open the dropdown
+  trigger: 'click' | 'hover';
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.types.ts:8]: Trigger should be optional and click by default.

File: src/blocks/dropdown/Dropdown.types.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { ReactNode } from 'react';
+
+export type DropdownProps = {
+  // This will be content upon clicking on which the dropdown overlay will open
+  children?: (isOpen: boolean) => ReactNode;
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.types.ts:5]:
```suggestion
  children: (isOpen: boolean) => ReactNode;
```
children should be mandatory for dropdown

File: src/blocks/dropdown/Dropdown.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        return 'bottomLeft';
+      case 'bottomRight':
+        return 'bottomRight';
+      default:
+        return 'bottom';
+    }
+  };
+
+  return (
+    <StyledDropdown
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:109]: Screenshot 2024-06-20 at 11 38 33 AM

@corlard3y 
The dropdown doesn't follows the correct figma design:https://www.figma.com/design/FvJL0JWHlsVohsHTERB1Lh/Push-Design-Foundations-[Future]?node-id=2663-23523&t=Hr7vGDBgMGSAlcxO-0

- Spacing between icon and label is off.
- There is no minimum width of the overlay item  as shown in figma. The overlay width is decided by the width of the label width.
-

Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:109]: Screenshot 2024-06-20 at 11 41 20 AM
Screenshot 2024-06-20 at 11 41 47 AM


Hover css is off

Comment (rohitmalhotra1420) [reply src/blocks/dropdown/Dropdown.tsx:109]: overlay is getting cut off in scroll container.

Screenshot 2024-06-20 at 11 38 33 AM

Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.tsx:109]: Screenshot 2024-06-20 at 11 43 04 AM

Overlay is not overlapping the scroll container and getting scrolled with the trigger element inside the scroll container.

**Ideally overlay should be a portal and should be hidden when the trigger element is out of viewport.**

File: src/blocks/menu/MenuItem.tsx [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+const MenuItem = forwardRef<HTMLElement, MenuProps>(({ icon, label, onClick, ...props }, ref) => {
+  const { mode } = useBlocksTheme();
+
+  return (
+    <StyledMenuItem
+      onClick={onClick}
+      mode={mode}
+      {...props}
+    >
+      {icon && <span className='menu-icon'>{icon()}</span>}
```
Comment (rohitmalhotra1420) [inline src/blocks/menu/MenuItem.tsx:52]:
```suggestion
      {icon && <span className='menu-icon'>{icon}</span>}
```

This is breaking at the moment.

Comment (rohitmalhotra1420) [reply src/blocks/menu/MenuItem.tsx:52]: - icon colors for light and dark mode are also off.
- icon size are are incorrect.
Screenshot 2024-06-20 at 11 44 37 AM
Screenshot 2024-06-20 at 11 45 59 AM

File: src/blocks/menu/MenuItem.tsx [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    }
+  }
+  .menu-label {
+    justify-content: center;
+    align-items: center;
+    color: ${({ mode }) => getBlocksColor(mode, { light: 'gray-1000', dark: 'gray-100' })};
+  }
+  cursor: pointer;
+  font-size: 15px;
+  /* Extra CSS props */
```
Comment (rohitmalhotra1420) [inline src/blocks/menu/MenuItem.tsx:39]: Use var(--s0) format for gap, margin and padding.
Use var(--r0) format for border-radius.
Use var(--pink-100) format for color and background-color.

File: src/blocks/dropdown/Dropdown.types.ts [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+  placement?: 'bottom' | 'bottomLeft' | 'top' | 'topRight' | 'bottomRight' | 'topLeft';
+  // on which action to open the dropdown
+  trigger: 'click' | 'hover';
+  // This is used for custom css instead of style prop, check Box/Text component
+  css?: FlattenSimpleInterpolation;
```
Comment (rohitmalhotra1420) [inline src/blocks/dropdown/Dropdown.types.ts:10]: FlattenSimpleInterpolation is not imported from styled-components.

File: src/blocks/menu/Menu.tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import { FC,forwardRef } from 'react';
```
Comment (rohitmalhotra1420) [inline src/blocks/menu/Menu.tsx:1]: Screenshot 2024-06-20 at 11 48 34 AM
Needs fixes

File: src/blocks/menu/Menu.types.ts [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import { CSSProperties, ReactNode } from 'react';
```
Comment (rohitmalhotra1420) [inline src/blocks/menu/Menu.types.ts:1]: Needs fixes
Screenshot 2024-06-20 at 11 49 25 AM

File: src/blocks/menu/MenuItem.tsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
+import { FC,forwardRef } from 'react';
```
Comment (rohitmalhotra1420) [inline src/blocks/menu/MenuItem.tsx:1]: Needs fixes
Screenshot 2024-06-20 at 11 49 50 AM

**Review by rohitmalhotra1420 — CHANGES_REQUESTED** [review]
Needs signficant amount of changes. Let's connect for this.

**Pull request closed without merging**